packages/search/src/util/merge-sort/merge.ts

```diff
@@ -0,0 +1,300 @@
+interface INode<T> {
+  data: T;
+  label: string;
+}
+
+interface IBinaryHeap<T> {
+  insert(node: INode<T>): void;
+  remove(): INode<T> | null;
+}
+
+type comparator<T> = (one: T, two: T) => number;
+
+type HeapDirection = "ASC" | "DESC";
+
+/**
+ * k-way merge implementation that merges (and potentially sorts) k individually sorted arrays
+ * based on the k-way merge algorithm (https://en.wikipedia.org/wiki/K-way_merge_algorithm)
+ *
+ * If a comparator function used to compare data is provided, a k-way merge sort is performed using
+ * a Binary Heap implementation. Otherwise data from each input array is merged incrementally
+ *
+ * The returned object contains merged (and potentially sorted) data, as well as key-value
+ * pairs that represent the index of a result set in the "data" parameter (the key) tied to how
+ * many results (the value) of that result set were added to the final merged data array.
+ * For example, if "data" represents an array of two result sets, with 3 and 5 results added to the final
+ * result set, respectively, the returned object would be { data: [...], "0": 3, "1": 5 }
+ *
+ * @param data An array of of result sets, each an array of type T
+ * @param resultLimit the maximum number of merged results to return, defaults to 10
+ * @param cmptr comparator function that takes in two instances of type T and returns a number to determine sorting order
```

Comment:
note what the rules are here regarding the number--i.e. +1 if a comes before b, -1 if b comes before a, 0 if equivalent, etc

Reply:
Good call, the name `comparator` is inspired by the corresponding [Java interface](https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html), but not everyone would know